src/ColorVectorSpace.jl

```diff
@@ -24,28 +21,75 @@ import Base:      conj, sin, cos, tan, sinh, cosh, tanh,
 
 MathTypes{T,C} = Union{AbstractRGB{T},TransparentRGB{C,T},AbstractGray{T},TransparentGray{C,T}}
 
-# convert(RGB{Float32}, NaN) doesn't and shouldn't work, so we need to reintroduce nan
+## Traits and key utilities
+
+# Return types for arithmetic operations
+multype(::Type{A}, ::Type{B}) where {A,B} = coltype(typeof(zero(A)*zero(B)))
+sumtype(::Type{A}, ::Type{B}) where {A,B} = coltype(typeof(zero(A)+zero(B)))
+divtype(::Type{A}, ::Type{B}) where {A,B} = coltype(typeof(zero(A)/oneunit(B)))
+powtype(::Type{A}, ::Type{B}) where {A,B} = coltype(typeof(zero(A)^zero(B)))
+multype(a::Colorant, b::Colorant) = coltype(multype(eltype(a),eltype(b)))
+sumtype(a::Colorant, b::Colorant) = coltype(sumtype(eltype(a),eltype(b)))
+divtype(a::Colorant, b::Colorant) = coltype(divtype(eltype(a),eltype(b)))
+powtype(a::Colorant, b::Colorant) = coltype(powtype(eltype(a),eltype(b)))
+
+coltype(::Type{T}) where {T<:Fractional} = T
+coltype(::Type{T}) where {T<:Number}     = floattype(T)
+
+acctype(::Type{T}) where {T<:FixedPoint}        = floattype(T)
+acctype(::Type{T}) where {T<:ShorterThanInt}    = Int
+acctype(::Type{Rational{T}}) where {T<:Integer} = typeof(zero(T)/oneunit(T))
+acctype(::Type{T}) where {T<:Real}              = T
+
+acctype(::Type{T1}, ::Type{T2}) where {T1,T2} = acctype(promote_type(T1, T2))
+
+acc(x::Number) = convert(acctype(typeof(x)), x)
+
+# Scalar binary RGB operations require the same RGB type for each element,
+# otherwise we don't know which to return
+color_rettype(::Type{A}, ::Type{B}) where {A<:AbstractRGB,B<:AbstractRGB} = _color_rettype(base_colorant_type(A), base_colorant_type(B))
+color_rettype(::Type{A}, ::Type{B}) where {A<:AbstractGray,B<:AbstractGray} = _color_rettype(base_colorant_type(A), base_colorant_type(B))
+color_rettype(::Type{A}, ::Type{B}) where {A<:TransparentRGB,B<:TransparentRGB} = _color_rettype(base_colorant_type(A), base_colorant_type(B))
+color_rettype(::Type{A}, ::Type{B}) where {A<:TransparentGray,B<:TransparentGray} = _color_rettype(base_colorant_type(A), base_colorant_type(B))
+_color_rettype(::Type{A}, ::Type{B}) where {A<:Colorant,B<:Colorant} = error("binary operation with $A and $B, return type is ambiguous")
+_color_rettype(::Type{C}, ::Type{C}) where {C<:Colorant} = C
+
+color_rettype(c1::Colorant, c2::Colorant) = color_rettype(typeof(c1), typeof(c2))
+
+arith_colorant_type(::C) where {C<:Colorant} = arith_colorant_type(C)
+arith_colorant_type(::Type{C}) where {C<:Colorant} = base_colorant_type(C)
+arith_colorant_type(::Type{Gray24}) = Gray
+arith_colorant_type(::Type{AGray32}) = AGray
+arith_colorant_type(::Type{RGB24}) = RGB
+arith_colorant_type(::Type{ARGB32}) = ARGB
+
+# Useful for leveraging iterator algorithms
+channels(c::AbstractGray)    = (gray(c),)
+channels(c::TransparentGray) = (gray(c), alpha(c))
+channels(c::AbstractRGB)     = (red(c), green(c), blue(c))
+channels(c::TransparentRGB)  = (red(c), green(c), blue(c), alpha(c))
```

Comment:
I think it's better to generalize these methods in ColorTypes as mentioned in https://github.com/JuliaGraphics/ColorTypes.jl/pull/184#issuecomment-644138321).

However, I think there is a need for more discussion as to whether `channels` should return an iterator or a tuple.
What about a constructor-style `Tuple` conversion method to "buy time" for the debate? It is obvious that `Tuple()` returns a tuple. On the other hand, I don't think it's good to overload `convert` for color --> tuple conversions, since it can be called implicitly.

Another trivial concern is that ColorTypes uses the term "component" (e.g. `comp1`) rather than "channel". I think the "channel" is appropriate for arrays of colors (i.e. images).

Reply:
I am not sure anymore that I want to export it. It's just a private method this way.

Reply:
If we introduce the concept of `echannels` different from `channels`, at least they should be documented. So I think they will be "semi-public", whether or not we export them.

Reply:
I don't think we should have both.